### ISSUE TITLE
GCP Adapter - catch exception and return status code 500 instead of 200.

### ISSUE THREAD

**Issue by saturnism** [spring-cloud/spring-cloud-function#539]
**Describe the bug**
When using GCP Adapter, if an exception was thrown, it doesn't not translate the exception to error status code. For example, if the function throws a runtime exceptions, it should translate into HTTP Status Code 500. Currently it's always translating to 

**Sample**
```
@Bean
  public Function<String, String> uppercase() {
    return value -> {
      throw new IllegalStateException("oops!");
    };
  
```

**Comment by olegz** [comment]
Why would you expect s-c-function to know anything about HTTP? 
The only module that is aware of it is `s-c-function-web` which in the context of this issue is identical to any adapter.

**Comment by saturnism** [comment]
the underlying functions framework is http based, and if an uncaught exception occurs, 200 response is not expected if there are errors, and non-2xx status codes would may trigger retries for the function.

**Comment by olegz** [comment]
I am not sure I follow. When you say the i am assuming you mean GCF?

**Comment by meltsufin** [comment]
@saturnism Do you think we should catch exceptions in [FunctionInvoker.service()](https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/main/java/org/springframework/cloud/function/adapter/gcp/FunctionInvoker.java#L87) and set the HTTP response code there?

**Comment by wwadge** [comment]
Exceptions shouldn't be used for flow control. How about we return a ResponseEntity instead like in MVC?

**Comment by saturnism** [comment]
it's not a matter of flow control, it's a simple contract that indicates if the function has a server side error, it should at least return status code 500, rather than 2xx.  Function in general do not have the knowledge of the underlying transport (HTTP, or not).

#540 discusses the ability to set a specific status code.

**Comment by saturnism** [comment]
> @saturnism Do you think we should catch exceptions in [FunctionInvoker.service()](https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/main/java/org/springframework/cloud/function/adapter/gcp/FunctionInvoker.java#L87) and set the HTTP response code there?

The underlying Function Framework already catches uncaught exceptions and return 500:
https://github.com/GoogleCloudPlatform/functions-framework-java/blob/master/invoker/core/src/main/java/com/google/cloud/functions/invoker/NewHttpFunctionExecutor.java

I'm actually surprised that through the s-c-function adapter, the exception did not result in the error.

**Comment by dzou** [comment]
Hey guys, I think #542 may be what we want. Basically it will catch the error thrown by the function in the adapter and if an exception is thrown, we will write `500` to the http response.

You can see the integration tests the effects of this change: https://github.com/spring-cloud/spring-cloud-function/pull/542/files#diff-6bc3016c8766b080ea84158f1d1048b9R66

See the usage of `TestRestTemplate` and the assertion on the response object.

**Comment by meltsufin** [comment]
@dzou Considering what @saturnism pointed to [in the invoker](https://github.com/GoogleCloudPlatform/functions-framework-java/blob/master/invoker/core/src/main/java/com/google/cloud/functions/invoker/NewHttpFunctionExecutor.java#L74), we shouldn't have to catch the exception because the functions invoker would, and it would set the status code. We must be missing something here.

**Comment by dzou** [comment]
Ah I see. What I observed is that Spring Cloud Function catches errors thrown by the function and will log these errors; consequently these errors never get propagated to the GCF Function Invoker.

I imagine the error throwing hierarchy to be:
```
Arrows point where the error is thrown:

User Function --> GCF Adapter --> Spring Cloud Function Framework (catches and logs as warning) --//--> GCF Invoker (The exception never reaches here)
```

**Comment by saturnism** [comment]
thnx @dzou @meltsufin for looking into this. do you know where in s-c-functions is doing the catch and log? i wonder if there is a simple configuration we can configure to make it catch, log, and rethrow.

**Comment by dzou** [comment]
Apologies I was wrong in what I said before.

The exception is being caught by the Google Cloud Functions Invoker, however it is unable to set the error response code properly because we call `.close()` on the http response in our adapter (The caller of our adapter which is GCF framework is unable to set the response code after we call close).

I'll send out the correct fix shortly.

**Comment by dzou** [comment]
Okay, this should be the correct fix: #544

**Comment by AndrewPonyk** [comment]
@dzou Your fix allows to 'throw' 500x errros. 

But for example if parameter of the function is 'not valid' would be nice to have possibility return some 400 response with meaningful error message

**Comment by saturnism** [comment]
> But for example if parameter of the function is 'not valid' would be nice to have possibility return some 400 response with meaningful error message

I think currently when marshaling fails, s-c-functions throws `com.google.gson.JsonSyntaxException`.   This can be intercepted and set the response code to `400` rather than propagating to to the functions framework to get a `500` error. 

Lastly, will it be possible to use `@Valid` to validate the input fields? If so, gcp adapter can also catch those exceptions.

However, I'm not sure if the entire exception can be propagated back to the caller. It's prob not a good practice because exception may contain sensitive information if un-careful.  Logging should be sufficient.

**Comment by meltsufin** [comment]
Fixed in #544.
